[mlir][OpenMP][flang] make private variable allocation implicit in omp.private

tblah opened the pull request:
The intention of this work is to give MLIR->LLVMIR conversion freedom to
control how the private variable is allocated so that it can be
allocated on the stack in ordinary cases or as part of a structure used
to give closure context for tasks which might outlive the current stack
frame. See RFC:
https://discourse.llvm.org/t/rfc-openmp-supporting-delayed-task-execution-with-firstprivate-variables/83084

For example, a privatizer for an integer used to look like
```mlir
  omp.private {type = private} @x.privatizer : !fir.ref<i32> alloc {
  ^bb0(%arg0: !fir.ref<i32>):
    %0 = ... allocate proper memory for the private clone ...
    omp.yield(%0 : !fir.ref<i32>)
  }
```

After this change, allocation become implicit in the operation:
```mlir
  omp.private {type = private} @x.privatizer : i32
```

For more complex types that require initialization after allocation, an init region can be used:
``` mlir
  omp.private {type = private} @x.privatizer : !some.type init {
  ^bb0(%arg0: !some.pointer<!some.type>, %arg1: !some.pointer<!some.type>):
    // initialize %arg1, using %arg0 as a mold for allocations
    omp.yield(%arg1 : !some.pointer<!some.type>)
  } dealloc {
    ^bb0(%arg0: !some.pointer<!some.type>):
    ... deallocate memory allocated by the init region ...
    omp.yield
  }
```

This patch lays the groundwork for delayed task execution but is not enough on its own.

After this patch all gfortran tests which previously passed still pass. There
are the following changes to the Fujitsu test suite:
- 0380_0009 and 0435_0009 are fixed
- 0688_0041 now fails at runtime. This patch is testing firstprivate variables with tasks. Previously we got lucky with the undefined behavior and won the race. After these changes we no longer get lucky. This patch lays the groundwork for a proper fix for this issue.

In flang the lowering re-uses the existing lowering used for reduction init and dealloc regions.

In flang, before this patch we hit a TODO with the same wording when
generating the copy region for firstprivate polymorphic variables. After
this patch the box-like fir.class is passed by reference into the copy
region, leading to a different path that didn't hit that old TODO but
the generated code still didn't work so I added a new TODO in
DataSharingProcessor.

---

Please read mlir changes first and then flang changes.

This has to be merged as one massive patch to avoid testsuite regressions,
but for ease of review I have broken it up into a series of (mostly) short patches.
In these patches I have only updated unit tests as I go along. They will only all
pass at the end of the series.

I have tried to make the changes to init/dealloc region generation self-contained
so that it is easier to see what changed when reviewing. For some commits I have
added extra commentary in the commit message for that commit.

By the end of this series the init/dealloc region generation is a bit hard to follow so I will refactor
it in a follow-up (hopefull NFC) patch.

In flang lowering I box up all arrays and pass the boxes by reference so
that the existing code for reduction init and dealloc regions can be
shared.

The TODOs for pointers, derived types, characters etc will be resolved
in later patches in this same series (to be squashed into this one). I
separated it to make it easier to review.

Assumed rank was already broken before this patch.
I can't find any mention of "assumed rank" in the openmp standard so
I guess it is not prohibited.

Other than the omp.private operation definition changes, the test changes
are mostly down to slightly different codegen from re-using the reduction
init region. That code is already well tested so I didn't want to change it.

@luporl  please could you review my re-implementation of your recent fix in 4d52dde46c2e121659415afa9d5cd0501aadb38c and especially the changes to the logic for the inclusion of barriers.

tblah committed:
[mlir][OpenMP] Change the definition of omp.private

The intention of this work is to give MLIR->LLVMIR conversion freedom to
control how the private variable is allocated so that it can be
allocated on the stack in ordinary cases or as part of a structure used
to give closure context for tasks which might outlive the current stack
frame. See RFC:
https://discourse.llvm.org/t/rfc-openmp-supporting-delayed-task-execution-with-firstprivate-variables/83084

In flang, before this patch we hit a TODO with the same wording when
generating the copy region for firstprivate polymorphic variables. After
this patch the box-like fir.class is passed by reference into the copy
region, leading to a different path that didn't hit that old TODO but
the generated code still didn't work so I added a new TODO in
DataSharingProcessor.

---

Please read mlir changes first and then flang changes.

In flang lowering I box up all arrays and pass the boxes by reference so
that the existing code for reduction init and dealloc regions can be
shared.

The TODOs for pointers, derived types, characters etc will be resolved
in later patches in this same series (to be squashed into this one). I
separated it to make it easier to review.

Assumed rank was already broken before this patch.
I can't find any mention of "assumed rank" in the openmp standard so
I guess it is not prohibited.

Other than the omp.private operation definition changes, the test changes
are mostly down to slightly different codegen from re-using the reduction
init region. That code is already well tested so I didn't want to change it.

tblah committed:
Support char types

The way boxchars are handled isn't ideal, but !fir.ref<!fir.boxchar<>>
seems to violate a lot of assumptions in the wider ecosystem of (hl)fir
helpers making it difficult to generate a copy region. I suspect
!fir.ref<!fir.boxchar<>> is not supposed to work (it looks like a
mutable character box, which isn't possible because a boxchar should be
an SSA value). Fixing this would be a big change beyond the scope of
this already too large PR.

tblah committed:
Support pointers

The copy region codegen hasn't changed as a result of this patch series.
However I think there is a bug in the copy region generated in
equivalence.f90. This patch series is already too big so I won't change
the copy region here.

tblah committed:
Fix bug when used with lastprivate

Arrays are boxed and those boxed arrays are passed by reference. This
can cause problems if we try to perform eager privatization (e.g.
lastprivate) on a variable already privatized using delayed
privatization. Work around this by loading the reference to the box.

tblah committed:
Support derived types

There are three cases handled:
- Boxed (maybe allocatable or pointer) scalar derived types
- Boxed (maybe allocatable or pointer) arrays of derived types
- Unboxed scalar derived types

Currently I support both boxed and unboxed derived types because unboxed
derived types aren't hlfir::Entities so I worry there could be cases
where they are in fact boxed.

The changes to parallel-private-clause.f90 are because of the arrays
becoming boxed. I re-organised the test a bit because the CHECK-DAGs
were matching completely the wrong lines outside of the parallel region.

tblah committed:
Implement lupori's derive type initialization patch

For reference see both
https://github.com/llvm/llvm-project/pull/120295 and
https://github.com/llvm/llvm-project/pull/121808

This changes the barrier logic.
See discussion here: https://github.com/llvm/llvm-project/pull/120295/files#r1910663473

tblah committed:
Fix compiler crash for unnesecarry len params

tblah committed:
Fix omp target maps for privatized symbols

tblah committed:
Deallocate derived types

This will deallocate in exactly the same cases as the old implementation
before this series. I expect some cases are missed but I want to avoid
functional changes wherever possible in this (already too big) patch
series.

tblah committed:
Fix crash lowering fir::EmboxOp whithout shape

This only affects POINTERs (which should not be initialized) and NULL
allocatables so the actual contents of the shape doesn't matter. This is
just so the embox operation is converted to LLVMIR correctly.

tblah committed:
Support arrays of polymorphic types

ergawy commented on flang/lib/Lower/OpenMP/PrivateReductionUtils.cpp:
```diff
@@ -96,17 +149,118 @@ fir::ShapeShiftOp Fortran::lower::omp::getShapeShift(fir::FirOpBuilder &builder,
+
+  if (!fir::isa_derived(derivedTy))
+    return;
+  assert(moldBox.getType() == newBox.getType());
```
nit: move to the function start to document the pre-conditions expected by it.

ergawy commented on flang/lib/Lower/OpenMP/PrivateReductionUtils.cpp:
```diff
@@ -96,17 +149,118 @@ fir::ShapeShiftOp Fortran::lower::omp::getShapeShift(fir::FirOpBuilder &builder,
+  if (const Fortran::semantics::DeclTypeSpec *declTypeSpec = sym.GetType())
+    if (const Fortran::semantics::DerivedTypeSpec *derivedTypeSpec =
+            declTypeSpec->AsDerived())
+      if (derivedTypeSpec->HasDefaultInitialization(
+              /*ignoreAllocatable=*/false, /*ignorePointer=*/true))
+        return true;
```
nit
```suggestion
      return derivedTypeSpec->HasDefaultInitialization(
              /*ignoreAllocatable=*/false, /*ignorePointer=*/true);
```

ergawy commented on flang/lib/Lower/OpenMP/PrivateReductionUtils.cpp:
```diff
@@ -96,17 +149,118 @@ fir::ShapeShiftOp Fortran::lower::omp::getShapeShift(fir::FirOpBuilder &builder,
+static mlir::Value generateZeroShapeForRank(fir::FirOpBuilder &builder,
+                                            mlir::Location loc,
+                                            mlir::Value moldArg) {
+  mlir::Type moldVal = fir::unwrapRefType(moldArg.getType());
```
```suggestion
  mlir::Type moldType = fir::unwrapRefType(moldArg.getType());
```

ergawy commented on flang/lib/Lower/OpenMP/PrivateReductionUtils.cpp:
```diff
@@ -148,41 +309,113 @@ void Fortran::lower::omp::populateByRefInitAndCleanupRegions(
+      mlir::Value null = builder.createNullConstant(loc, boxTy.getEleTy());
+      mlir::Value nullBox;
+      if (shape)
+        nullBox = builder.create<fir::EmboxOp>(
+            loc, boxTy, null, shape, /*slice=*/mlir::Value{}, lenParams);
```
I think we can get rid of the `else`, right? If `shape` is an empty value, it is the same as passing `Value{}` I think.

tblah commented:
I'm not sure why the bot didn't run on this.

@llvm/pr-subscribers-flang-openmp

ergawy commented on flang/lib/Optimizer/OpenMP/MapsForPrivatizedSymbols.cpp:
```diff
@@ -55,15 +55,19 @@ class MapsForPrivatizedSymbolsPass
-    assert(declOp &&
-           "Expected defining Op of privatized var to be hlfir.declare");
+    assert(definingOp &&
+           "Privatizing a block argument without any hlfir.declare");
```
If we don't restrict ourselves to `DeclareOp`, do we still need that assertion? If so, the error message needs to be generalized.

tblah replied:
MLIR values can come from two places:
1. the result of an operation
2. a block argument

We can't assume that getting the defining operation produces a non-null result because the value might be a block argument.

For example,
```
func.func @func(%arg0 : !type0, %arg1 : !type1) {
  %0 = hlfir.declare %arg0
  omp.private(%0, %arg1)
}
```
Here we can get a defining operation for `%0` because it is the `hlfir.declare` for that function argument (which is the normal way flang would lower a function argument). `%arg1` has no defining operation because it is a block argument.

tblah replied:
I was nervous to make any functional change to the target stuff because I don't know how to test it. The previous implementation also wouldn't have worked for block arguments.

I can fix this if somebody at AMD is willing to test `omp target` for me?

ergawy replied:
We can remove the assertion and see if any of our smoke tests triggers any problems (I will monitor that). Then we can add handle whatever is missing and add a regression test.

tblah replied:
Thank you! [44bbbe8](https://github.com/llvm/llvm-project/pull/124019/commits/44bbbe83e44e1cbfc7f32779920dc86325773f7b)

ergawy commented on flang/lib/Optimizer/OpenMP/MapsForPrivatizedSymbols.cpp:
```diff
@@ -55,15 +55,19 @@ class MapsForPrivatizedSymbolsPass
-    Value varPtr = declOp.getBase();
+    // Some types are boxed immediately before privatization. These have other
+    // operations in between the privatization and the declaration. It is safe
+    // to use var directly here because they will be boxed anyay.
```
```suggestion
    // to use var directly here because they will be boxed anyway.
```

ergawy reviewed (commented):
Thanks for the great work @tblah! I partially reviewed since this is a huge PR (understandably). I will come back and continue reviewing the rest later.

tblah commented on mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:
```diff
@@ -34,52 +34,48 @@ def PrivateClauseOp : OpenMP_Op<"private", [IsolatedFromAbove, RecipeInterface]>
+    alloca) variable is passed to the initialization region which does everything
+    else (e.g. initialization of Fortran runtime descriptors). Information about
+    how to initialize the copy from the original item should be given in the
+    copy region, and if needed, how to deallocate memory (allocated by the
+    initialization region) in the dealloc region.
```
Thanks for reviewing the PR. These readability comments are very helpful.

The LLVM lowering is responsible for managing the implicitly allocated memory (it is usually just a stack allocation).

The `dealloc` region undoes what is done in the `init` region. So for example for a `!fir.box<!fir.heap<!fir.array<?xi32>>>`, the box is allocated implicitly on the stack and then the `init` region performs the heap allocation for the array and the dealloc region frees that heap allocation.

I will add an example in the omp.private documentation to make it clearer.

tblah commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -488,44 +559,34 @@ void DataSharingProcessor::doPrivatize(const semantics::Symbol *sym,
+      // TODO: currently there are false positives from dead uses of the mold
+      // arg
+      if (!result.getInitMoldArg().getUses().empty())
+        mightHaveReadMoldArg = true;
```
Quite a lot of cases actually. Some examples:
- Derived types that need a runtime initialization call but which have no allocatable components (the allocatable component initialization does read from the mold argument)
- Pointers only need to be set to NULL
- Arrays with compile-time known sizes. In this case the box is allocated implicitly but memory for the actual array needs to be allocated (on the heap) in the init region and the box has to be set up to point to that allocated array and contain the correct shape.
- Characters with compile-time known sizes

Currently there are some cases where the there are dead loads from the mold argument or maybe even reading character length parameters which never get used. These are eventually removed by DCE but would lead to a false positive here. I hope to improve the init region generation in a future patch to remove these cases.

tblah committed:
Readability improvments

tblah commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -488,44 +559,34 @@ void DataSharingProcessor::doPrivatize(const semantics::Symbol *sym,
-      firOpBuilder.create<mlir::omp::YieldOp>(hsb.getAddr().getLoc(),
-                                              yieldedValue);
+    // Populate the `init` region.
+    const bool needsInitialization =
```
Done in [7cb83e3](https://github.com/llvm/llvm-project/pull/124019/commits/7cb83e335cfd25ab750e8c15c951dfb9198ac7a1)

ergawy reviewed (approved):
LGTM! Thanks Tom! However, I have to admit, the `fir` dialect type system is still a "semi-"blackbox to me, so someone more familiar with it needs to carefully review changes in `PrivateReductionUtils.cpp`.

luporl reviewed (commented):
Thanks for all the effort you have put into this @tblah. I'm looking forward to privatization support in delayed tasks.

I like your approach for detecting the need of barriers, based on the uses of `moldArg`, it's clever.

I have a few comments, but no major concerns with this PR.

tblah committed:
Fix MLIR test

I made a mistake when rewriting the privatizer, which led to the test
reading incorrectly.

tblah commented on mlir/test/Target/LLVMIR/openmp-private.mlir:
```diff
@@ -125,22 +123,23 @@ llvm.func @parallel_op_private_multi_block(%arg0: !llvm.ptr) {
-  %c1 = llvm.mlir.constant(1 : i32) : i32
-  %0 = llvm.alloca %c1 x f32 : (i32) -> !llvm.ptr
-  llvm.br ^bb1(%arg0, %0 : !llvm.ptr, !llvm.ptr)
+// Check that the body of the parallel region loads from the private clone.
+// CHECK: [[PAR_REG2]]:
+// CHECK:        %{{.*}} = load float, ptr %[[ORIG_PTR3]], align 4
```
Good spot. I made a mistake when I rewrote the privatizer leading it to yield the wrong value.

tblah replied:
Fixed in https://github.com/llvm/llvm-project/pull/124019/commits/4325873687bb18cd9000a12b369e593667c99f29

tblah committed:
More readability improvments

tblah commented on flang/lib/Lower/OpenMP/PrivateReductionUtils.cpp:
```diff
@@ -371,13 +426,15 @@ void Fortran::lower::omp::populateByRefInitAndCleanupRegions(
-    }
+    builder.setInsertionPointToStart(initBlock);
+    mlir::Type boxedTy = fir::BoxType::get(ty);
+    mlir::Value newBox =
+        builder.create<fir::EmboxOp>(loc, boxedTy, allocatedPrivVarArg);
+    mlir::Value moldBox = builder.create<fir::EmboxOp>(loc, boxedTy, moldArg);
+    initializeIfDerivedTypeBox(
+        builder, loc, newBox, moldBox, needsInitialization,
+        /*isFirstPrivate=*/kind == DeclOperationKind::FirstPrivate);
```
Agreed. It is even worse than just derived types, we load `moldArg` to fetch the length parameters whether or not there really are any. That load (and probably these embox operations) get removed by DCE but not before the barriers are created.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1338,15 +1338,14 @@ findAssociatedValue(Value privateVar, llvm::IRBuilderBase &builder,
-                    llvm::SmallVectorImpl<llvm::Value *> &llvmPrivateVars,
-                    const llvm::OpenMPIRBuilder::InsertPointTy &allocaIP,
-                    llvm::DenseMap<Value, Value> *mappedPrivateVars = nullptr) {
+static llvm::Expected<llvm::BasicBlock *> allocateAndInitPrivateVars(
```
I updated it in https://github.com/llvm/llvm-project/pull/124019/commits/ca35ec29e81dfd3222d4ebba91937c96b5080f91

luporl commented on flang/test/Lower/OpenMP/derived-type-allocatable.f90:
```diff
@@ -13,32 +13,42 @@ module m1
-!CHECK-NEXT:  omp.yield
+!CHECK:       omp.yield
 !CHECK:       } dealloc {
 !CHECK:       fir.call @_FortranAAllocatableDeallocate
 !CHECK:       omp.yield
 
 !CHECK-LABEL: omp.private {type = firstprivate} @_QMm1Ftest_array
+!CHECK:       fir.call @_FortranAInitialize(
 !CHECK-NOT:   fir.call @_FortranAInitializeClone
 !CHECK:       omp.yield
 
 !CHECK-LABEL: omp.private {type = private} @_QMm1Ftest_array
+!CHECK:       fir.call @_FortranAInitialize(
 !CHECK:       fir.call @_FortranAInitializeClone
-!CHECK-NEXT:  omp.yield
+!CHECK:       omp.yield
 
 !CHECK-LABEL: omp.private {type = private} @_QMm1Ftest_scalar
 !CHECK:       fir.call @_FortranAInitializeClone
-!CHECK-NEXT:  omp.yield
+!CHECK:       omp.yield
```
My intention with these `CHECK-NEXT`s was to check that no extra barriers were added.

To keep this check and make the intention more explicit, it could be changed to:
```
CHECK-NOT: omp.barrier
CHECK: omp.yield
```

tblah replied:
I've added CHECK-NOT lines here: [ce45103](https://github.com/llvm/llvm-project/pull/124019/commits/ce451030c7df6e75285710dc83b2fc447041e713)

tblah committed:
Allow block arguments in MapsForPrivatizedSymbols

tblah committed:
Add negative tests for barriers

tblah committed:
Fix test broken after rebase

tblah commented:
@luporl ping for review

luporl reviewed (approved):
LGTM, thanks.

If I understood correctly, with this change the generated code will have more barriers than before, although some of those might be fixing existing race conditions.
Is my understanding correct, that the goal is to prioritize correctness for now and to remove unnecessary barriers in the future?

tblah commented:
> If I understood correctly, with this change the generated code will have more barriers than before, although some of those might be fixing existing race conditions. Is my understanding correct, that the goal is to prioritize correctness for now and to remove unnecessary barriers in the future?

Yes that is my intention

luporl commented on flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:
```diff
@@ -184,7 +218,8 @@ bool DataSharingProcessor::needBarrier() {
     if (sym->test(semantics::Symbol::Flag::OmpLastPrivate) &&
-        (sym->test(semantics::Symbol::Flag::OmpFirstPrivate) || callsInitClone))
+        (sym->test(semantics::Symbol::Flag::OmpFirstPrivate) ||
+         mightHaveReadHostSym))
```
AFAIU, a barrier should not be needed in this case, since it is not possible to have concurrent read/write of the same memory region, as `i` is only written to in the last iteration and, as it seems, `x` is only read during privatization (this could probably be optimized, since `x`'s shape is known at compile time).

I hadn't noticed it before, but having a single `mightHaveReadHostSym` variable for all privatized symbols is a problem indeed. A possible solution would be to have a per-symbol `mightHaveReadHostSym`.

skatrak replied:
Thank you for the quick response. I just created #127074 to address the second part of the issue. Though it looks like, from what you're saying, we might also want to treat induction variables differently as well?

luporl replied:
Thanks for the quick fix.
I only wanted to point out that, in this case, `x` wouldn't even need to be read at all.
Actually, this optimization is implemented in #125901.

Do you think there is any advantage in treating induction variables differently?

tblah replied:
Sorry for my slow response @skatrak, I was on holiday. Thanks for fixing it, and thanks @luporl for chiming in.
